$$ {\huge \text{ Disciplina: Lógica Fuzzy }}\\  \\  \text{ Cap 9 - Tomada de Decisão em Sistemas Fuzzy }\\
\ \  \ \ \ \ 
\\
 { \it  \text{ Discente: Prof. Dr. João Fernando }  }
 \\
\text{ Monitor: Arnaldo Araújo} \\
\text{ Mestrado e Doutorado em Ciências da Computação - UFC }$$


# Exemplo 9.1 - NOS LIVROS ANTIGOS É O CAPÍTULO 10

<img src="Images/exem1011.png">

<img src="Images/exem1012.png">

In [5]:
import numpy as np
import skfuzzy as fuzz
import skfuzzy.control as ctrl
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.lines as mlines
from matplotlib.font_manager import FontProperties
%matplotlib inline 

In [8]:
# Score de 0.4 para performance, 0.3 para custo, 0.2 para disponibilidade, 0.1 para software
w = np.array([0.4,0.3,0.2,0.1])
w

array([ 0.4,  0.3,  0.2,  0.1])

In [10]:
R = np.array([[0.1,0.3,0.4,0.2],[0.,0.1,0.8,0.1],[0.1,0.6,0.2,0.1],[0.1,0.4,0.3,0.2]])
R

array([[ 0.1,  0.3,  0.4,  0.2],
       [ 0. ,  0.1,  0.8,  0.1],
       [ 0.1,  0.6,  0.2,  0.1],
       [ 0.1,  0.4,  0.3,  0.2]])

In [11]:
e = fuzz.maxmin_composition(w,R)
e

array([[ 0.1,  0.3,  0.4,  0.2]])

$$\large{\text{Logo, temos que o produto será enquadrado à categoria ADEQUADO }}$$

# Exemplo 9.2

<img src="Images/exem1021.png">

In [13]:
I1 = np.array([1,0.8])# 3,7
I2 = np.array([0.7,1.]) # 4,6
I3 = np.array([0.8,1.,0.5]) # 2,4,8

$$I1 \geqslant
I2 $$

In [22]:
I1I2 = max(min(I1[1],I2[0]),min(I1[1],I2[1]))
I1I2

0.80000000000000004

$$\large{\text{Similarmente para todas as combinações }}$$

<img src="Images/exem1022.png">

$$\large{\text{Logo: }}$$

<img src="Images/exem1023.png">

$$\large{\text{Deste modo:  } I2 \geqslant
I1  \geqslant I3  }$$

# Exemplo 9.3

<img src="Images/exem1031.png">

<img src="Images/exem1032.png">

In [25]:
Mat = np.array([[1.,0.5,0.3,0.2],[0.7,1.,0.8,0.9],[0.5,0.3,1.,0.7],[0.3,0.1,0.3,1.]])


array([[ 1. ,  0.5,  0.3,  0.2],
       [ 0.7,  1. ,  0.8,  0.9],
       [ 0.5,  0.3,  1. ,  0.7],
       [ 0.3,  0.1,  0.3,  1. ]])

In [32]:
#Matriz Condicional
def MaCcond(A):
    n = A.shape[0]
    C = np.zeros((n,n))
    for i in range(n):
        for j in range(n):
            C[i,j] = A[j,i] / max(A[j,i],A[i,j])
    return C

#Vetor das ordens
def V_Ordem(M):
    n = M.shape[0]
    v = np.zeros((n))
    for i in range(n):
        v[i] = min(M[i,:])
    return v
            

In [34]:
Fcond = MaCcond(Mat)
Fcond

array([[ 1.        ,  1.        ,  1.        ,  1.        ],
       [ 0.71428571,  1.        ,  0.375     ,  0.11111111],
       [ 0.6       ,  1.        ,  1.        ,  0.42857143],
       [ 0.66666667,  1.        ,  1.        ,  1.        ]])

In [36]:
posicao = V_Ordem(Fcond)
posicao

array([ 1.        ,  0.11111111,  0.42857143,  0.66666667])

Deste modo, temos que  X1 é o primeiro da ordem diantes do valor 1, 
X4 é o segundo com pontuação de 0.666, X3 o terceiro com pontuação de 0.428 e, por fim, X2 com pontuação de 0.111.


$$\Large{\text{ X1 >
 X4 >
 X3 >
 X2  }}$$

# Exemplo 9.5 - TOMADA DE DECISÃO MULTIOBJETIVO

<img src="Images/exem1051.png">

<img src="Images/exem1052.png">

In [47]:
O = np.array([[0.4,1.,0.1],[0.7,.8,0.4],[0.2,0.4,1],[1.0,0.5,0.5]])
b = np.array([0.8,.9,0.7,0.5])
b_barra =  1-b


In [50]:
min(b)

0.5

In [49]:
O

array([[ 0.4,  1. ,  0.1],
       [ 0.7,  0.8,  0.4],
       [ 0.2,  0.4,  1. ],
       [ 1. ,  0.5,  0.5]])

In [87]:
def Escolha(O,b_barra):
    M = np.copy(O)
    l,c = O.shape
    v = np.zeros((c,1))
    for j in range(c):
        com1 = np.zeros((l,1))
        for i in range(l):
            com1[i] = max(M[i,j],b_barra[i])
            menor = min(com1)
        v[j] = menor
    D = np.argmax(v)
    return D+1,v
            



In [88]:
Opcao,v = Escolha(O,b_barra)
Opcao

2

In [89]:
#Posição 1 do vetor.
v

array([[ 0.3],
       [ 0.4],
       [ 0.2]])

$$\large{\text{Logo, a opção escolhida é a segunda (MURO DE CONCRETO)}}$$

## CENÁRIO 2

In [97]:
#NOVO B
b2 =  np.array([0.5,0.7,0.8,0.7])
b2_barra = 1-b2
Opcao2,v2 = Escolha(O,b2_barra)


In [98]:
Opcao2

2

$$\large\text{TEMOS UM EMPATE}$$

In [99]:
v2

array([[ 0.2],
       [ 0.4],
       [ 0.4]])

$$\large\text{Você remove o fator que causou o empate}$$

In [101]:
Dc2 = min(max( 0.5, 1), max( 0.3, 0.8 ),max(0.3, 0.5))
Dc2

0.5

In [103]:
Dc3 = min(max( 0.5, 0.1), max( 0.2, 1. ),max(0.3, 0.5))
Dc3

0.5

$$\large\text{Você remove novamente o fator que causou o empate}$$

In [105]:
DDc2= min(max( 0.5, 1), max( 0.3, 0.8 ))
DDc2

0.8

In [108]:
DDc3 =  max( 0.2, 1. ) 
DDc3

1.0

$$\large\text{Logo a opção 3 será a escolhida - GABION WALL}$$

## Exercícios

<img src="Images/exer92.png">

<img src="Images/exer93.png">

<img src="Images/exer932.png">

<img src="Images/exer97.png">